framework/src/vectorpostprocessors/SamplerBase.C

```diff
@@ -44,104 +44,106 @@ void
-    _comm.allgather(_values_tmp[i], false);
-
-  // Next... figure out the correct sorted positions of each value
-  std::vector<size_t> sorted_indices;
-
-  switch (_sort_by)
-  {
-    case 0: // x
-      Moose::indirectSort(_x_tmp.begin(), _x_tmp.end(), sorted_indices);
-      break;
-    case 1: // y
-      Moose::indirectSort(_y_tmp.begin(), _y_tmp.end(), sorted_indices);
-      break;
-    case 2: // z
-      Moose::indirectSort(_z_tmp.begin(), _z_tmp.end(), sorted_indices);
-      break;
-    case 3: // id
-      Moose::indirectSort(_id_tmp.begin(), _id_tmp.end(), sorted_indices);
-      break;
-  }
-
-  // Use the sorted_indices to copy out of the temporary vectors and into the actual output vectors
-  for (unsigned int i=0; i<_variable_names.size(); i++)
+  /**
+   * We have several vectors that all need to be processed in the same way.
+   * Rather than enumerate them all, let's just create a vector a pointers
+   * and work on them that way.
+   */
+  // Initialize the pointer vector with the position and ID vectors since we sort only on these.
+  std::vector<VectorPostprocessorValue *> vec_ptrs { { &_x, &_y, &_z, &_id } };
+  std::vector<std::size_t> sorted_indices;
+
+  // Gather up each of the partial vectors then choose one to sort by
+  for (auto i = beginIndex(vec_ptrs); i < vec_ptrs.size(); ++i)
   {
-    _values[i]->resize(sorted_indices.size());
+    _comm.allgather(*vec_ptrs[i], false);
 
-    for (unsigned int j=0; j<sorted_indices.size(); j++)
-      (*_values[i])[j] = _values_tmp[i][sorted_indices[j]];
+    // Sort by the vector chosen by the user
+    if (i == _sort_by)
+      Moose::indirectSort(vec_ptrs[i]->begin(), vec_ptrs[i]->end(), sorted_indices);
   }
 
-  _x.resize(sorted_indices.size());
-  _y.resize(sorted_indices.size());
-  _z.resize(sorted_indices.size());
-  _id.resize(sorted_indices.size());
-
-  for (unsigned int i=0; i<sorted_indices.size(); i++)
+  // Now gather the contents of each of the sample value vectors
+  for (auto i = beginIndex(_variable_names); i < _variable_names.size(); ++i)
+    _comm.allgather(*_values[i], false);
+
+  /**
+   * We now have one sorted vector. The remaining vectors need to be sorted according to that vector.
+   * We'll need a temporary vector to hold values as we map them according to the sorted indices.
+   * After that, we'll swap the vector contents with the sorted vector to get the values
+   * back into the original vector.
+   */
+  // This vector is used as temp storage to sort each of the remaining vectors according to the first
+  auto vector_length = sorted_indices.size();
+  VectorPostprocessorValue tmp_vector(vector_length);
+
+  // Here we extend the vector by all the remaining values vector before performing the final sort
+  vec_ptrs.insert(vec_ptrs.end(), _values.begin(), _values.end());
+
+#ifndef NDEBUG
+  for (const auto vec_ptr : vec_ptrs)
+    if (vec_ptr->size() != vector_length)
+      mooseError("Vector length mismatch");
```

Comment:
Yeah - it would nice to be able to actually handle this... like only sort the values that exist or something.  But it's an edge case that can't happen with any of our existing objects that use this interface right now.